src/tqec/circuit/detector/construction_test.py

```diff
@@ -0,0 +1,163 @@
+                    new_measure_moment.append(
+                        stim.CircuitInstruction("M", inst.targets_copy())
+                    )
+                    new_reset_moment.append(
+                        stim.CircuitInstruction("R", inst.targets_copy())
+                    )
+                elif inst.name in ["MRX", "MRY"]:
+                    new_measure_moment.append(
+                        stim.CircuitInstruction(f"M{inst.name[2]}", inst.targets_copy())
+                    )
+                    new_reset_moment.append(
+                        stim.CircuitInstruction(inst.name[1:], inst.targets_copy())
+                    )
+                else:
+                    new_measure_moment.append(inst)
+            new_circuit += new_measure_moment
+            if new_measure_moment[-1].name != "TICK":
+                new_circuit.append(stim.CircuitInstruction("TICK", []))
+            new_circuit += new_reset_moment
+            new_circuit.append(stim.CircuitInstruction("TICK", []))
+        else:
+            new_circuit += moment
+    return new_circuit
+
+
+def prepend_qubit_coords_for_repetition_code(circuit: stim.Circuit) -> stim.Circuit:
+    new_circuit = stim.Circuit()
+    for i in range(circuit.num_qubits):
+        new_circuit.append(stim.CircuitInstruction("QUBIT_COORDS", [i], [i]))
+    new_circuit += circuit
+    return new_circuit
+
+
+def load_extra_test_cases() -> list[tuple[str, stim.Circuit]]:
+    cases = []
+    circuits_dir = pathlib.Path(__file__).parent / "extra_test_circuits"
+    for path in circuits_dir.glob("*.stim"):
+        circuit_id = path.stem
+        circuit = stim.Circuit.from_file(path)
+        cases.append((circuit_id, circuit))
+    return cases
+
+
+def gen_test_cases() -> list[tuple[str, stim.Circuit]]:
+    cases = []
+    code_tasks = [
+        # "repetition_code:memory",  # NOT WORKING FOR ROUNDS=1 or 2
+        "surface_code:rotated_memory_x",
+        "surface_code:rotated_memory_z",
+        "surface_code:unrotated_memory_x",
+        "surface_code:unrotated_memory_z",
+    ]
+    for code_task, distance, rounds, decompose_MR in itertools.product(
+        code_tasks, (3, 5), (1, 2, 3, 11, 15), (True, False)
+    ):
+        circuit_id = f"{code_task}:d{distance}:r{rounds}:decomposeMR={decompose_MR}"
+        circuit = reorder_detectors_for_equality_check(
+            stim.Circuit.generated(
+                code_task=code_task,
+                distance=distance,
+                rounds=rounds,
+                after_clifford_depolarization=0.001,
+                before_round_data_depolarization=0.005,
+                after_reset_flip_probability=0.01,
+                before_measure_flip_probability=0.01,
+            )
+        )
+        if code_task.startswith("repetition_code"):
+            circuit = prepend_qubit_coords_for_repetition_code(circuit)
+        if decompose_MR:
+            circuit = replace_MR_with_M_then_R(circuit)
```

Comment:
Why do you test both cases (keeping `MR` and replacing them with `M` followed by `R`)?

I am trying to enforce explicitly some invariants on the dataclasses, and ended-up with that for the `Fragment` class:
```py
@dataclass(frozen=True)
class Fragment:
    """A sub-circuit guaranteed to span the locations between two nearest collapsing
    moments or those outside the error detection regions.

    Attributes:
        circuit: the circuit represented by the instance.
        end_stabilizer_sources: sources (reset instructions) that generate the
            stabilizers that might be found at the end of the circuit.
        begin_stabilizer_sources: sources (measurement instructions) that generate
            the stabilizers that might be found at the beginning of the circuit.
        sources_for_next_fragment: sources (reset instructions) that generate
            stabilizers after the measurements of this Fragment instance and so
            should be accounted for in the next Fragment.
    """

    circuit: stim.Circuit
    end_stabilizer_sources: list[PauliString] = field(default_factory=list)
    begin_stabilizer_sources: list[PauliString] = field(default_factory=list)
    sources_for_next_fragment: list[PauliString] = field(default_factory=list)

    def __post_init__(self):
        moments = list(
            circ.copy() if isinstance(circ, stim.Circuit) else circ
            for circ in iter_stim_circuit_by_moments(self.circuit)
        )
        print(f"Moments: {moments}")
        if not moments:
            raise TQECException("Empty circuit found in Fragment.")
        if any(isinstance(m, stim.CircuitRepeatBlock) for m in moments):
            raise TQECException(
                "Found a stim.CircuitRepeatBlock in a Fragment instance. "
                "Please use FragmentLoop."
            )
        # Check that the first moment of the fragment contains reset operations.
        if not has_reset(moments[0]):  # type: ignore we already checked that there is no instance of stim.CircuitRepeatBlock
            print(moments)
            raise TQECException("First moment of Fragment not containing any reset.")
        # Check that the last moment of the fragment contains measurement operations.
        if not has_measurement(moments[-1]):  # type: ignore we already checked that there is no instance of stim.CircuitRepeatBlock
            print(moments)
            raise TQECException(
                "Last moment of Fragment not containing any measurement."
            )
```

Basically, for any `Fragment`, it seems to me that we expect the first moment of the stored circuit to contain reset operations, and the last moment to contain measurement operations. But replacing `MR` with `M` + `R` breaks that assumption.

Is my constraint incorrect, in which case what is the invariant for `Fragment`, or is the constraint correct, in which case the test case seems incorrect?

Reply:
I finally understood. There is no real invariant on the `circuit` attribute of `Fragment` (except that it is not supposed to contain an instance of `stim.CircuitRepeatBlock`). Everything is encoded in the other attributes, and so even if the circuit does not start with resets, or does not strictly end with measurements, this is still fine.